### ISSUE TITLE
Add-HPOVBaseline consumes/leaks memory (PS on CentOS)

### ISSUE THREAD

**Issue by frippe75** [HewlettPackard/POSH-HPOneView#427]
# Expected Behavior

 Add-HPOVBaseline -File <file> should add the baseline to OV without consuming all available memory

### Actual Behavior

```PowerShell
PS /home/frta/> Add-HPOVBaseline -File ./spp-2019.09.0.iso
Add-HPOVBaseline : Exception of type 'System.OutOfMemoryException' was thrown.                                                              At line:1 char:1                                                                                                                            
+ Add-HPOVBaseline -File ./spp-2019.09.0.iso                                                                                                
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                
+ CategoryInfo          : NotSpecified: (:) [Add-HPOVBaseline], OutOfMemoryException                                                        
+ FullyQualifiedErrorId : OutOfMemoryException,Add-HPOVBaseline 
```

Running a parallel free command every second shows that memory is consumed at the same pase as the upload

```Bash
[frta@docker ]$ free -m
              total        used        free      shared  buff/cache   available
Mem:           3789        1218        2145          26         425        2212
Swap:          3983        3600         383
[frta@docker ]$ free -m
              total        used        free      shared  buff/cache   available
Mem:           3789        1333        2019          27         436        2096
Swap:          3983        3584         399
[frta@docker ]$ free -m
              total        used        free      shared  buff/cache   available
Mem:           3789        1885        1135          27         769        1544
Swap:          3983        3574         409
[frta@docker ]$ free -m
              total        used        free      shared  buff/cache   available
Mem:           3789        2145         622          27        1021        1284
Swap:          3983        3574         409
[frta@docker]$ free -m
              total        used        free      shared  buff/cache   available
Mem:           3789        2435         300          27        1053         993
Swap:          3983        3574         409
[frta@docker]$ free -m
              total        used        free      shared  buff/cache   available
Mem:           3789        2682         120          27         987         747
Swap:          3983        3576         407
```

### Version Information

```PowerShell
PS /home/frta/> Get-HPOVVersion -ApplianceVer

synergy-demo-500.lab.net            LibraryVersion Path
------------------------            -------------- ----
ApplianceVersion: 5.00.00.400525.00 5.0.2226.2609  /home/frta/.local/share/powershell/Modules/HPOneView.500/5.0.2226.2609

PS /home/frta/> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.2.3
PSEdition                      Core
GitCommitId                    6.2.3
OS                             Linux 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by ChrisLynchHPE** [comment]
Not sure this is an issue with the library.  When uploading the file, a 4MB buffer is used.  It does not increase, as the same buffer is used each time.  Nor is the entire SPP ISO loaded into memory.  This might be an artifact of .NetCore, or the OS itself. 

How much RAM did you allocate to your CentOS VM?

**Comment by frippe75** [comment]
I increased memory to allow for the full file to fit and ran again.
Now I'm bumping into a "Streams too long" at 2GB.

Could be a known issue https://github.com/PowerShell/PowerShell/issues/4129. It's not an exact match but Invoke-RestMethod seems to fail at the same place as the Upload-File code does and probably has a similar issue with underlying functioncalls/interfaces. Invoke-WebRequest seems to work. The code for uploading multipart files is really tight. Was planning on writing a smaller Upload-File2 function in the POSH library and calling that from Add-HPOVBaseline. 

If I have some time I will look into Invoke-RestMethod/Invoke-WebRequest.

Commenting out the following lines masks the issue
```PowerShell
# $rs.write($readbuffer, 0, $byteCount)
# $rs.flush()
```
Noticed you set the `$uploadRequest.AllowWriteStreamBuffering = $false` ... Feels like that part is not working...

**Comment by ChrisLynchHPE** [comment]
Hmm...  This appears to happen with large SPP ISO's only.  The upload function inside has used the process for a very long time.  In regression testing, I never ran into this when uploading sample ISO's, which have been smaller than 2GB in size.  This is reproducible in PowerShell 6 and PowerShellCore 6, on any platform.  So I'll have to look into this further.

**Comment by frippe75** [comment]
Feels like the Stream type in could be of the type MemoryStream: I think that was the issue faced in my linked issue above. 2GB seems to be the largest object size unless you complile Powershell using "AllowLargeObjects". So trying to avoid getting into the memory in the first place is the way to go.

**Comment by ChrisLynchHPE** [comment]
>  largest object size unless you complile Powershell using "AllowLargeObjects"

No, you have that wrong.  While yes PowerShell _could_ be compiled with that property set, the comment from the Microsoft engineer was targeting the library or module being used, not PowerShell itself. 

After some testing, this is looking like an issue with .NetCore itself and either the `MemoryStream` class, or `Stream` base class.  I saw that when using PowerShell 5.x, the process would ballon, but not at the sharp and consistent degree as PowerShell 6.  I did see that the `FileStreamReader` class implements a `Flush()` method, and using it within PowerShell 5.x helps reduce memory usage.  But it did nothing for PowerShell 6.  I'll open an issue, or look to see if there is one with .NetCore on this topic, as it is only happening with PowerShell/PowerShellCore 6.

**Comment by frippe75** [comment]
Maybe look at the Invoke-WebRequest implementation which contains uploading with progress... And increase the iso used for testing. But that's a given I guess. Would have liked to dive into this myself but limited on time. Cheers!

**Comment by ChrisLynchHPE** [comment]
Unfortunately, I cannot use Invoke-WebRequest as a replacement to the C# class method currently being used.  And as I stated, this only happens with ISO's greater than 2GB in size.  The more I look into the root cause, the more it appears to be a bug with .NetCore, not the library.  I'll be reaching out to the .NetCore team on this topic in the near future.

**Comment by DungKHoang** [comment]
Hey Chris
Any update on this issue?
I was not able to upload a full ISO to OV with PSCore on Ubuntu
Thanks for any update

**Comment by ChrisLynchHPE** [comment]
No because this is a DotNetCore issue.  ~~The only recommendation I can provide is to add more RAM to the Ubuntu image (if a VM).~~

This recommendation is no longer viable.  Only leaving here for others to see it brought up as a suggestion, but does not work as a solution.

**Comment by ChrisLynchHPE** [comment]
To follow up on this, I have recently observed severe latency issues with `System.Net.HttpWebRequest` even with PowerShell 7 and DotNetCore 3.  I'm working on replacing the core functionality of the internal `Download-File` and `Upload-File` helper functions with `System.Net.HttpClient`.  This doesn't mean the rest of the library will start supporting async HTTP operations, but it is kind of a step in that direction.

**Comment by shashi-kalyan** [comment]
Hey chris,

I am also facing issue when uploading 7GB with SUSE linux . Can you let me know any update on this. I tried increasing RAM but not resolved.

**Comment by ChrisLynchHPE** [comment]
I am still trying to find an appropriate solution here that isn't to use `Invoke-WebRequest` or `Invoke-RestMethod`.

**Comment by ChrisLynchHPE** [comment]
I have implemented a `-UseInvokeWebRequest` switch parameter to the library that provides a workaround.  The downside is that transfer progress is not displayed.  This was implemented in the 6.60, 8.60, 9.00 and 9.10 libraries.
